core/src/main/java/org/apache/iceberg/DefaultValueParser.java

```diff
@@ -0,0 +1,190 @@
+        return ByteBuffer.allocate(((Types.FixedType) type).length()).put(fixedBytes);
+      case BINARY:
+        Preconditions.checkArgument(defaultValue.isTextual(),
+            "Cannot parse %s to a %s value", defaultValue, type);
+        byte[] binaryBytes = BaseEncoding.base16().decode(defaultValue.textValue().toUpperCase(Locale.ROOT));
+        return ByteBuffer.wrap(binaryBytes);
+      case LIST:
+        Preconditions.checkArgument(defaultValue.isArray(),
+            "Cannot parse %s to a %s value", defaultValue, type);
+        List<Object> defaultList = Lists.newArrayList();
+        for (JsonNode element : defaultValue) {
+          defaultList.add(parseDefaultFromJson(type.asListType().elementType(), element));
+        }
+        return defaultList;
+      case MAP:
+        Preconditions.checkArgument(
+            defaultValue.isObject() && defaultValue.has("keys") && defaultValue.has("values") &&
+                defaultValue.get("keys").isArray() && defaultValue.get("values").isArray(),
+            "Cannot parse %s to a %s value",
+            defaultValue,
+            type);
+        Map<Object, Object> defaultMap = Maps.newHashMap();
+        JsonNode keys = defaultValue.get("keys");
+        JsonNode values = defaultValue.get("values");
+        List<JsonNode> keyList = Lists.newArrayList(keys.iterator());
+        List<JsonNode> valueList = Lists.newArrayList(values.iterator());
+
+        for (int i = 0; i < keyList.size(); i++) {
+          defaultMap.put(
+              parseDefaultFromJson(type.asMapType().keyType(), keyList.get(i)),
+              parseDefaultFromJson(type.asMapType().valueType(), valueList.get(i)));
+        }
+        return defaultMap;
+      case STRUCT:
+        Preconditions.checkArgument(defaultValue.isObject(),
+            "Cannot parse %s to a %s value", defaultValue, type);
+        Map<Integer, Object> defaultStruct = Maps.newHashMap();
+        for (Types.NestedField subField : type.asStructType().fields()) {
+          String fieldIdAsString = String.valueOf(subField.fieldId());
+          Object value = defaultValue.has(fieldIdAsString) ? parseDefaultFromJson(
+              subField.type(),
+              defaultValue.get(fieldIdAsString)) : null;
+          if (value != null) {
+            defaultStruct.put(subField.fieldId(), value);
+          }
+        }
+        return defaultStruct;
+      default:
+        return null;
+    }
+  }
+
+  public static Object convertJavaDefaultForSerialization(Type type, Object value) {
+    switch (type.typeId()) {
+      case DATE:
+        return DateTimeUtil.formatEpochDays((int) value);
+      case TIME:
+        return DateTimeUtil.formatTimeOfDayMicros((long) value);
+      case TIMESTAMP:
+        return DateTimeUtil.formatEpochTimeMicros((long) value, ((Types.TimestampType) type).shouldAdjustToUTC());
+      case FIXED:
+      case BINARY:
+        return BaseEncoding.base16().encode(((ByteBuffer) value).array());
```

Comment:
This is an incorrect and unsafe use of `array()`:
* The buffer may not be backed by `byte[]`
* The buffer may use an `arrayOffset()`
* The buffer may have a non-zero position
* The buffer may have a limit before the end of the array

You can use `ByteBuffers.toByteArray(value)` if you need `byte[]`, although that may copy in some cases. I think it should be fine here since most of the time the backing array can be used directly because it will be created by `fromJson`.

Reply:
Sure, done. I think the `byte[]` will always come from the `fromJson` 's `wrap()` call.

Reply:
@rzhang10, yes, as I said it should be fine "most of the time".

The problem is that you have no guarantee about the byte buffer that is passed in. Hence this is unsafe.